refactor: split app code into several libraries

shoom3301 opened the pull request:
# Summary

First of all, sorry for this huge PR. It turned out that `modules/wallet` is highly coupled with the app code.
Most of the dependencies:
 - legacy/utils
 - legacy/const
  - legacy/hooks
  - legacy/assets
 - common/pure
 - legacy/state
 
 Initially, I just converted these directories into libs and then moved back some specific files that are coupled with the app code.
I know, it doesn't fit to the main idea of this monorepo, libs should be based on domains. However, in the current situation, it required giant efforts to refactor `legacy` folder in order to split it into domains.
During the refactoring a few times I copypasted some interfaces like `Nullish` or `Writeable` and consts like `ONE_FRACTION` to avoid `someLib -> app` dependencies. Probably, in the future I should move them into `libs/core`.

Some of the hooks that moved into libs were based on `useSingleCallResult` which is a part of uniswap multicall. Uniswap multicall is based on redux and it's highly coupled with app code. And actually, there is no reason to use `useSingleCallResult`, I guess uniswap devs did it to avoid excessive request, actually it's a goal of SWR.

I moved a few Updater into `common/updaters` from `legacy/state` folder because I wanted to extract `legacy/state` into a lib, [but later I realized that it's not possible](https://github.com/cowprotocol/cowswap/pull/3109/commits/0510aab324a2e35ea606ef7329719fab0ff40c08). The legacy/state code is concretely coupled with app and can't be separated. Anyway, those updaters are left in `common/updaters`, probably I should move them back.

## The refactoring steps:
1. Just git-moved `apps/cowswap-frontend/src/modules/wallet` into `libs/wallet/src` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/17bf7bb249c2f3c67b9cbac6986bc1b1f532d881))
2. Just git-moved `legacy/hooks` into `libs/common-hooks` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/034a6adbb90811658c16d3a0038330f2c326e569))
3. NX lib boilerplate for `libs/wallet` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/20d27ef38b915b159becceb715dc8ef2c96830c0))
4. NX lib boilerplate for `libs/common-hooks` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/555b36aa0dba34a4505ffae7ef3fca83d43eaef4))
5. NX lib boilerplate for `libs/common-utils` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/b49f5fb78d55a32790e5a9604f288564fd7d21de))
6. Just git-moved `legacy/utils` into `libs/common-utils` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/0f3f41478caebdc927c112c99ab9076ce6839e38))
7.  Just git-moved `legacy/assets` into `libs/assets` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/23cb2a07039de11ea0dee727fb785512da086a80))
8. Just git-moved `legacy/utils` into `libs/common-const` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/571c7bb58d113be85b775cdb9ada9628f00de4b5))
9. Git-moved `lib/utils` into `libs/common-utils` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/08de481edab62ec87845f6822c4a37797cb363e5))
10. After moving files into libs I started fixing their imports and exports from index.ts. Pretty often I had to fixing default imports. ([example](https://github.com/cowprotocol/cowswap/pull/3109/commits/2ec9898ee0c75c9b5f7a8d1b4917ee0fbc922d26))
11. NX lib boilerplate for `libs/ens` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/4f2be53987e899a5213e49ce6fd5b8e0632862d7))
12. Moved ENS hooks into `libs/ens` and replaced `useSingleCallResult` with `useSWR` to avoid dependencies from multicall.
13. Git-moved `Button, FiatAmount, Loader, Row, SelectDropdown, TokenAmount, TokenSymbol` components from `common/pure` into `libs/ui` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/0b8cc1e3ca138a9611a38addcd1c528646cb3209))
14. Got rind of default export `store` from `legacy/state` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/26deb98cf0b411d302199297ed77c27e65da6957))
15. Moved a few updaters into `common/updaters` ([1](https://github.com/cowprotocol/cowswap/pull/3109/commits/7ba29231005e71b687067936a465f2553af99162), [2](https://github.com/cowprotocol/cowswap/pull/3109/commits/f7127b88522dc255c654e11ff3416da49254aca0), [3](https://github.com/cowprotocol/cowswap/pull/3109/commits/f460b70964ab27f1f9d928e8eecfd37d8a76b471), [4](https://github.com/cowprotocol/cowswap/pull/3109/commits/0f9acc4a962ae96d32d455b23508dadfc0e20d0c), [5](https://github.com/cowprotocol/cowswap/pull/3109/commits/07dc3e81703c5113bc5016d76a4d36dab454b858))
16. NX lib boilerplate for `libs/analytics` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/2ec00d291d9fd3ded297a49467469bfcefbbfe73))
17. Just git-moved `legacy/components/analytics` into `libs/analytics` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/0cce53a339ab0b3aff9515a43db3352513ca50ed))
18. Moved `ExternalLink` components to UI lib ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/a39a614e8fd9ca34d15220ab6efdedce5a79456e))
19. Wallet connections components were linked to `useIsActiveWallet` which is based on redux. To get rid of this dependency I refactored `ConnectWalletOptions` and other components. ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/970ad6c012b517fbcd3a0d162233d4e2e4e302c9))
20. NX lib boilerplate for `libs/core` ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/2a4fabe40c85e561d486cab13691928d9b3c29fa))
21. Moved `jotaiStore` and `gasPirceAtom` into `libs/core`. Probably it's not the best solution, however, those entities are used in `libs/wallet` and I need to move them in some new place that will be independent from app code ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/b83dafb7545c393b305cfe0b6b30ed5e0d9580c3#diff-ca1d8d131dd38ebde3026f5b356756e1c8f528f9fe6a6f860936a6a301ddde6f))
22. Moved `api/gnosisSafe` to `libs/core` with the same goals as above ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/f619a4d5796794f1a119d7b6e16939e16fe2f480))
23. Removed `@lingui/macro` usage from `libs/wallet` and replaced `styled-components/macro` -> `styled-components` in `libs/ui` because there were problems with the build and I haven't found any solutions for it. I know, it's not the best solution, but I don't see any other ways right now ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/ab6de08d05deff53b5d433cb878700e36542d9a7))
24. I've had to cut some corners regarding UI enum usage. Since it is declared in the app files I can't import if from libraries. So, in libs I replaced it by vars hardcode: `var(${UI.COLOR_CONTAINER_BG_01}) => var(--cow-container-bg-01)`. To avoid expanding this PR I propose to keep this hack, because later we can easily move the enum in some common lib and replace imports and hardcoded values. ([commit](https://github.com/cowprotocol/cowswap/pull/3109/commits/d97e1b713647bc0921f0b2f27c1a77d67dbafc2a))

  # To Test

1. Actually, everything :)
2. ENS name
3. ENS Avatar
4. All supported wallets
5. Google analytics
6. Dark theme

shoom3301 commented on libs/common-hooks/src/useColor/index.ts:
```diff
@@ -1 +1,55 @@
-export * from './useColorMod'
+import { useLayoutEffect, useState } from 'react'
```
Got rid of Mod file

shoom3301 commented on libs/wallet/src/web3-react/hooks/useIsSmartContractWallet.ts:
```diff
@@ -0,0 +1,87 @@
+  return isSmartContractWallet
+}
+
+function useIsArgentWallet(): boolean {
```
Moved useIsArgentWallet and useIsAmbireWallet into this file because they are used only here

shoom3301 commented on libs/ens/src/hooks/useENSAddress.ts:
```diff
@@ -1,28 +1,21 @@
  * Does a lookup for an ENS name to find its address.
  */
 export function useENSAddress(ensName?: string | null): { loading: boolean; address: string | null } {
-  const debouncedName = useDebounce(ensName, 200)
```
Removed debounce from `useENSAddress`, it looks redundant

shoom3301 commented on libs/ens/src/hooks/useENSName.ts:
```diff
@@ -12,12 +11,11 @@ import { useENSResolverMethod } from './useENSResolverMethod'
  * Note this is not the same as looking up an ENS name to find an address.
  */
 export function useENSName(address?: string): { ENSName: string | null; loading: boolean } {
-  const debouncedAddress = useDebounce(address, 200)
```
Removed debounce from `useENSName`

shoom3301 commented on apps/cowswap-frontend/src/legacy/state/lists/hooks.ts:
```diff
@@ -104,20 +99,7 @@ export function useCombinedActiveList(): TokenAddressMap {
 // list of tokens not supported on interface for various reasons, used to show warnings and prevent swaps and adds
 export function useUnsupportedTokenList(): TokenAddressMap {
-  // get hard-coded broken tokens
-  const brokenListMap = useMemo(() => tokensToChainTokenMap(BROKEN_LIST), [])
```
Actually, BROKEN_LIST and UNSUPPORTED_TOKEN_LIST are empty and we don't need them

shoom3301 commented on apps/cowswap-frontend/src/modules/application/containers/App/Updaters.tsx:
```diff
@@ -37,7 +38,8 @@ export function Updaters() {
       <UserUpdater />
       <ApplicationUpdater />
-      <EnhancedTransactionUpdater />
+      <FinalizeTxUpdater />
```
Split `EnhancedTransactionUpdater` into to independent updaters

shoom3301 commented on libs/wallet/src/api/utils/connection.ts:
```diff
@@ -1,5 +1,4 @@
 import CoinbaseWalletIcon from '../assets/coinbase.svg'
-import FortmaticIcon from '../assets/formatic.png'
```
Removed FORTMATIC because it's not used

shoom3301 commented on libs/wallet/src/web3-react/containers/Web3Status/index.tsx:
```diff
@@ -29,16 +27,14 @@ export function Web3Status() {
 
   const latestProvider = localStorage.getItem(STORAGE_KEY_LAST_PROVIDER)
 
-  const { pendingActivity } = useCategorizeRecentActivity()
```
Got rid of `useCategorizeRecentActivity` dependency

alfetopito commented:
EOA wallet (Metamask):

- [ ] Swap
- [ ] Approve + swap
- [ ] Wrap + approve + swap (eth-flow)
- [ ] Wrap/unwrap
- [ ] EthFlow swap
- [ ] EthFlow cancellation
- [ ] Order placement and fill (Limit/Twap)
- [ ] Hard cancel order (Swap/Limit/Twap)
- [ ] Soft cancel order (Swap/Limit/Twap)

alfetopito commented:
Account page:

- [ ] add token to Metamask
- [ ] approve token
- [ ] COW balance displaying
- [ ] token search by name/address
- [ ] favorite token add/remove

Common (WalletConnect/Metamask):

- [ ] network switching
- [ ] network persisting between modules
- [ ] wallet disconnect
- [ ] change wallet
- [ ] balances loading
- [ ] appzi
- [ ] all FAQ pages open

alfetopito commented:
WalletConnect v2:

- [ ] Wrap/unwrap on SWAP
- [ ] Wrap/unwrap on LIMIT
- [ ] Wrap/unwrap on TWAP
- [ ] Just swap
- [ ] Approve + swap
- [ ] Wrap + approve + swap (eth-flow)
- [ ] EthFlow cancellation
- [ ] Order placement and fill (LIMIT)
- [ ] Hard cancel order (SWAP/LIMIT)
- [ ] Soft cancel order (SWAP/LIMIT)

Smart contract wallet (Safe):

- [ ] Just swap
- [ ] Approve + swap
- [ ] Wrap + approve + swap (eth-flow)
- [ ] Wrap/unwrap
- [ ] Order placement and fill (Limit/Twap)
- [ ] Hard cancel order (Swap/Limit/Twap)

alfetopito commented:
In the Safe, the balances took a long time to be updated.
I had 0 BAT and after [my order completed](https://dev.explorer.cow.fi/goerli/orders/0x318ad460b84ec193e2d654bf367ff9ab822647e069f97b8d6b643fbb2795dd395895fcf0a96a4d19e17d2dd759926d1a02781ef8650ab846?tab=overview) took close to a minute to see it updated in the UI
![image](https://github.com/cowprotocol/cowswap/assets/43217/fb9dc2aa-ddf1-414e-b4fe-9fce5b107c27)

anxolin reviewed (commented):
Since this is a big PR i plan to give multiple reviews for ti as I review different aspects of it. 

For now I just read the description and the commits pointed out there, I have a general idea. 

In general I'm super happy with the changes, it feels like an intermediate PR because i don't agree with many of the new modules. But they were also not good in the legacy dir. 

If this gets merged, i'd love to continue shapping this. 

One concern though. The app took like 4 times more to build than usual.  I know that having modules will accelerate in principle, but is strange that the build of a web app takes more than 4min

Also, when i built a second time, it FAILED. It was fast but the result was giving a weird error: 

`mv: rename dist/apps/cowswap-frontend/* to build/*: No such file or directory`

anxolin commented:
This is the error i mentioned above:

![image](https://github.com/cowprotocol/cowswap/assets/2352112/9e0a1c38-7c2e-45e8-b85a-3a657fb73492)

anxolin commented:
Related to the error above, this is what i see in my local:

![image](https://github.com/cowprotocol/cowswap/assets/2352112/5e825a6c-1fca-419d-b7d7-7f3006b69f18)

anxolin commented:
running `nx build cowswap` fixed the problem:

![image](https://github.com/cowprotocol/cowswap/assets/2352112/85e30378-d869-4909-8774-0c9d15a041d3)


![image](https://github.com/cowprotocol/cowswap/assets/2352112/bb7e438f-8d15-458e-9df0-a55e8502c1c4)



But after this, it keeps failing if I do a `yarn build`, see:
![image](https://github.com/cowprotocol/cowswap/assets/2352112/a4614bc0-c319-42a8-90c4-c3067dd33be3)

anxolin commented:
@shoom3301 something i noticed is, you used `@cowswap` namespace for the new libraries. 

However we don't own that namespace. we own `@cowprotocol`

Shouldn't we use this one so we can decide to publish to NPM if we want?

alfetopito commented:
Done with testing, nothing else to report

shoom3301 commented:
https://github.com/cowprotocol/cowswap/pull/3109#issuecomment-1727262198

@alfetopito I wasn't able to reproduce it :( But I noticed a significant difference between Metamask and Safe. In Safe it takes 2x more time to update balances.
For balances we use `multicall` from `@uniswap/redux-multicall` and as far as I understand, it re-fetches only when the current block number increases. Probably it was some temporary problem with Safe.

shoom3301 commented:
@anxolin thank you!
I've fixed `postbuild`, just replaced `mv` by `cp`.
When nx hit cache it doesn't create artifacts again and it's a reason why the directory is empty.

shoom3301 commented:
I also changed `nx/tasks-runners/default` to `nx-cloud` to get better caching.
Now we hit a cache more often and in this case CI build takes 1.5 min.

shoom3301 commented:
> @anxolin thank you! I've fixed `postbuild`, just replaced `mv` by `cp`. When nx hit the cache it doesn't create artifacts again and it's a reason why the directory is empty.

@anxolin Actually, we don't even need this `postbuild` script, removed it.

alfetopito reviewed (approved):
Approved!
Be aware I have not fully retested it.
Only checked out, built locally and checked the page loaded fine.
Build was SUUUUPER fast BTW 👏

anxolin commented:
13s!!!! 🎉

all caches working as a charm!

<img width="1739" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/a6502432-c9fc-4e95-8ca2-383cd888fe11">

anxolin reviewed (approved):
Approved! didn't test the flows, but now it uses all the caches from NX cloud! good from my side as logn as the flows have been minimally tested
